pkg/timeutils/utils.go

```diff
@@ -1,36 +1,102 @@
 // as the given reference time minus the amount of the duration.
-func GetTimestamp(value string, reference time.Time) string {
+func GetTimestamp(value string, reference time.Time) (string, error) {
 	if d, err := time.ParseDuration(value); value != "0" && err == nil {
-		return strconv.FormatInt(reference.Add(-d).Unix(), 10)
+		return strconv.FormatInt(reference.Add(-d).Unix(), 10), nil
 	}
 
 	var format string
+	var parseInLocation bool
+
+	// if the string has a Z or a + or three dashes use parse otherwise use parseinlocation
+	parseInLocation = !(strings.ContainsAny(value, "zZ+") || strings.Count(value, "-") == 3)
+
 	if strings.Contains(value, ".") {
-		format = time.RFC3339Nano
+		if parseInLocation {
+			format = RFC3339NanoLocal
+		} else {
+			format = time.RFC3339Nano
+		}
+	} else if strings.Contains(value, "T") {
+		if parseInLocation {
+			format = RFC3339Local
+		} else {
+			format = time.RFC3339
+		}
+	} else if parseInLocation {
+		format = DateLocal
 	} else {
-		format = time.RFC3339
+		format = DateWithZone
 	}
 
-	loc := time.FixedZone(time.Now().Zone())
-	if len(value) < len(format) {
-		format = format[:len(value)]
+	var t time.Time
+	var err error
+
+	if parseInLocation {
+		t, err = time.ParseInLocation(format, value, time.FixedZone(time.Now().Zone()))
+	} else {
+		t, err = time.Parse(format, value)
 	}
-	t, err := time.ParseInLocation(format, value, loc)
+
 	if err != nil {
-		return value
+		// if there is a `-` then its an RFC3339 like timestamp otherwise assume unixtimestamp
+		if strings.Contains(value, "-") {
+			return "", err // was probably an RFC3339 like timestamp but the parser failed with an error
+		}
+		return value, nil // unixtimestamp in and out case
+	}
+
+	return fmt.Sprintf("%d.%09d", t.Unix(), int64(t.Nanosecond())), nil
+}
+
+// ParseTimestamps returns seconds and nanoseconds from a timestamp that has the
+// format "%d.%09d", time.Unix(), int64(time.Nanosecond()))
+// if the incoming nanosecond portion is longer or shorter than 9 digits it is
+// converted to nanoseconds.  The expectation is that the seconds and
+// seconds will be used to create a time variable.  For example:
+//     seconds, nanoseconds, err := ParseTimestamp("1136073600.000000001",0)
+//     if err == nil since := time.Unix(seconds, nanoseconds)
+// returns seconds as def(aultSeconds) if value == ""
+func ParseTimestamps(value string, def int64) (int64, int64, error) {
+	if value != "" {
+		sa := strings.SplitN(value, ".", 2)
+		s, err := strconv.ParseInt(sa[0], 10, 64)
+		if err != nil {
+			return s, 0, err
+		}
+		if len(sa) == 2 {
+			n, err := strconv.ParseInt(sa[1], 10, 64)
+			if err != nil {
+				return s, n, err
+			}
+			// should already be in nanoseconds but just in case convert n to nanoseonds
+			n *= int64(math.Pow(float64(10), float64(9-len(sa[1]))))
+			return s, n, nil
+		}
+		return s, 0, nil
 	}
-	return strconv.FormatInt(t.Unix(), 10)
+	return def, 0, nil
```

Comment:
just a minor nit, you can avoid a layer of indentation by doing something like:

```
if value == "" {
 return def, 0, nil
}
sa := strings.SplitN(value, ".", 2)
...
```

https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow